backends/kv/inferschema.go

```diff
@@ -35,57 +39,140 @@ func (b *Backend) inferSchema(request *frames.ExecRequest) error {
 	if err != nil {
 		return err
 	}
 
-	rowSet := []map[string]interface{}{}
-	indicies := []string{}
+	var rowSet []map[string]interface{}
+	var keys []string
 
 	for rowNum := 0; rowNum < maxrec && iter.Next(); rowNum++ {
 		row := iter.GetFields()
 		rowSet = append(rowSet, row)
-		index, ok := row["__name"]
+		key, ok := row["__name"]
 		if !ok {
 			return fmt.Errorf("key (__name) was not found in row")
 		}
-		indicies = append(indicies, index.(string))
+		keys = append(keys, key.(string))
 	}
 
 	if iter.Err() != nil {
 		return iter.Err()
 	}
 
-	labels := map[string]interface{}{}
-	frame, err := frames.NewFrameFromRows(rowSet, indicies, labels)
+	newSchema, err := schemaFromKeys(keyField, rowSet)
 	if err != nil {
-		return fmt.Errorf("Failed to create frame - %v", err)
+		return err
 	}
 
 	nullSchema := v3ioutils.NewSchema(keyField, "")
-	newSchema := v3ioutils.NewSchema(keyField, "")
 
-	for _, name := range frame.Names() {
-		col, err := frame.Column(name)
-		if err != nil {
-			return err
+	return nullSchema.UpdateSchema(container, table, newSchema)
+}
+
+func schemaFromKeys(keyField string, rowSet []map[string]interface{}) (v3ioutils.V3ioSchema, error) {
+	columnNameToValue := make(map[string]interface{})
+	columnCanBeFullKey := make(map[string]bool)
+	columnCanBePrimaryKey := make(map[string]bool)
+	columnCanBeSortingKey := make(map[string]bool)
+
+	for _, row := range rowSet {
+		keyValue := row["__name"].(string)
+		var primaryKeyValue string
+		var sortingKeyValue string
+		indexOfDot := strings.Index(keyValue, ".")
+		if indexOfDot >= 0 && indexOfDot < len(keyValue)-1 {
+			sortingKeyValue = keyValue[indexOfDot+1:]
+			primaryKeyValue = keyValue[:indexOfDot]
+		}
+		for attrName, attrValue := range row {
+			if attrName == "__name" {
+				continue
+			}
+			previousValue, ok := columnNameToValue[attrName]
+			if ok {
+				previousType := reflect.TypeOf(previousValue)
+				currentType := reflect.TypeOf(attrValue)
+				if previousType != currentType {
+					return nil, errors.Errorf("Type %v of %v did not match type %v of %v for column %s.", previousType, previousValue, currentType, attrValue, attrName)
+				}
+			}
+			columnNameToValue[attrName] = attrValue
+			if _, ok = columnCanBeFullKey[attrName]; !ok {
+				columnCanBeFullKey[attrName] = true
+			}
+			columnCanBeFullKey[attrName] = columnCanBeFullKey[attrName] && attrValue == keyValue
+			if primaryKeyValue != "" {
+				if _, ok = columnCanBePrimaryKey[attrName]; !ok {
+					columnCanBePrimaryKey[attrName] = true
+				}
+				columnCanBePrimaryKey[attrName] = columnCanBePrimaryKey[attrName] && attrValue == primaryKeyValue
+			}
+			if sortingKeyValue != "" {
+				if _, ok = columnCanBeSortingKey[attrName]; !ok {
+					columnCanBeSortingKey[attrName] = true
+				}
+				columnCanBeSortingKey[attrName] = columnCanBeSortingKey[attrName] && attrValue == sortingKeyValue
+			}
+		}
+	}
+
+	var primaryKeyField string
+	var sortingKeyField string
+	if keyField == "" {
```

Comment:
if keyField != "" then no need to do infer, you can check in the start and skip the entire logic

Reply:
Still need to infer to get the type of each column.